api/include/opentelemetry/trace/propagation/HttpTraceContext.cc

```diff
@@ -0,0 +1,183 @@
+    // it would make sense to define those keys in a single file only and had to be referenced by other files to store and retrieve values,
+    // otherwise I will not be able to access any fields, for example, "current-span" as CreateKey("current-span") will
+    // not work because the id is different when the value is put into despite the Key is also created from
+    // CreateKey("current-span").
+    // Don't know if I get the correct understanding there.
+    new_values.setValue(Context.SPAN_KEY,span);
+    return new_values;
+}
+
+static Span GetCurrentSpan(Context &context) {
+    Span span = get_value(Context.SPAN_KEY, context);
+    if (span == NULL) {
+        return NULL;
+    }
+    return span;
+}
+
+// The HttpTraceContext provides methods to extract and inject
+// context into headers of HTTP requests with traces.
+// Example:
+//    HttpTraceContext.inject(setter,&carrier,&context);
+//    HttpTraceContext.extract(getter,&carrier,&context);
+class HttpTraceContext : public HTTPTextFormat
+{
+    public:
+        List<nostd::string_view> fields() {
+            static const auto* FIELDS = new std::vector<nostd::string_view>({TRACE_PARENT, TRACE_STATE});
+            return FIELDS;
+        }
+
+        void inject(Setter setter, T &carrier, const Context &context) override {
+            common::AttributeValue span = GetCurrentSpan(context);
+            if (span == NULL || !span.getContext().isValid()) {
+                // We don't have span.getContext() in span.h, should we just use span? As well as acquiring validity. (I do know how to implement them though)
+                return;
+            }
+            injectImpl(setter, carrier, span.getContext());
+        }
+
+        Context extract(Getter getter, const T &carrier, Context &context) override {
+            SpanContext spanContext = extractImpl(carrier, getter);
+            return SetSpanInContext(trace.DefaultSpan(spanContext), context);
+        }
+
+    private:
+        static const nostd::string_view TRACE_PARENT = "traceparent";
+        static const nostd::string_view TRACE_STATE = "tracestate";
+        static const int VERSION_BYTES = 2;
+        static const int TRACE_ID_BYTES = 32;
+        static const int PARENT_ID_BYTES = 16;
+        static const int TRACE_FLAG_BYTES = 2;
+        static const int TRACE_DELIMITER_BYTES = 3;
+        static const int HEADER_SIZE = VERSION_BYTES + TRACE_ID_BYTES + PARENT_ID_BYTES + TRACE_FLAG_BYTES + TRACE_DELIMITER_BYTES;
+        static const int TRACESTATE_MAX_MEMBERS = 32;
+        static const nostd::string_view TRACESTATE_KEY_VALUE_DELIMITER = "=";
+        static const std::regex TRACESTATE_ENTRY_DELIMITER_SPLIT_PATTERN("[ \t]*,[ \t]*");
+
+        static Context checkNotNull(Context &arg, nostd::string_view errorMessage) {
+            if (arg == NULL) {
+                throw new NullPointerException("error: null " + errorMessage);
```

Comment:
Please use a macro instead of a throw, so that the implementation can operate without exceptions.

Reply:
I am actually not sure about this, should I be returning back a constant Context with some indicator of it being invalid?